ENH: Add fpspace function

settle opened the pull request:
ENH: Add fpspace function

Hybrid of linspace and logspace based on floating-point space.

TST: Added tests for fpspace

settle committed:
ENH: Add fpspace function

Hybrid of linspace and geomspace based on floating-point space.  Useful when
more precision is needed around a reference point (often zero), and less
precision is needed when moving further away from the reference point.  Provides
greater dynamic range than linspace (though less than geomspace) while
maintaining equal or better precision than linspace nearby the reference point
(and even more so compared to geomspace).  Often a desired scale for plotting
concentrated data but with a wide range due to long tails (major ticks at
geomspaced intervals and minor ticks at linspaced intervals), or in a lossy
compression of integers into fewer bits.

TST: Added tests for fpspace

bashtage commented on numpy/core/function_base.py:
```diff
@@ -356,3 +356,126 @@ def geomspace(start, stop, num=50, endpoint=True, dtype=None):
+        The base of the floating-point space.
+        Default is 2.0.
+    exponentbits : integer, optional
+        Number of exponent bits to in floating-point space.
```
To in

bashtage commented on numpy/core/function_base.py:
```diff
@@ -356,3 +356,126 @@ def geomspace(start, stop, num=50, endpoint=True, dtype=None):
+
+    if refpoint is None:
+        refpoint = 0.5 * (start + stop)
+    elif refpoint < start or stop < refpoint:
```
Is this check right? Elif stop < refpoint < start?

charris commented:
Please provide a more extended explanation of this function and tell us why it is needed.

settle commented:
ENH: Add fpspace function

Hybrid of linspace and geomspace based on floating-point space.  Useful when
more precision is needed around a reference point (often zero), and less
precision is needed when moving further away from the reference point.  Provides
greater dynamic range than linspace (though less than geomspace) while
maintaining equal or better precision than linspace nearby the reference point
(and even more so compared to geomspace).  Often a desired scale for plotting
concentrated data but with a wide range due to long tails (major ticks at
geomspaced intervals and minor ticks at linspaced intervals), or in a lossy
compression of integers into fewer bits.

TST: Added tests for fpspace

settle commented:
I provided more detailed parameter and behavior explanations, fixed the typo above, removed the above constraint on the reference point being bound by the interval, added support for both increasing and decreasing intervals and endpoint behavior like in the current spacing functions.

Sorry, still trying getting the hang of git and GitHub pull requests.

settle commented:
I'm not quite sure what to make on the continuous-integration build failures..., I followed as close as possible the linspace and geomspace function code in the same files so I'm not catching why those functions pass (actually, do they pass?) but fpspace fails.

shoyer commented on numpy/core/function_base.py:
```diff
@@ -356,3 +356,153 @@ def geomspace(start, stop, num=50, endpoint=True, dtype=None):
+    u_start = 0
+    u_stop = _nx.power(base, significandbits) - 1
+    u_num = u_stop - u_start + 1
+    u = linspace(u_start, u_stop, num=u_num, endpoint=True, dtype=dt)
```
If you look at your test failures in Travis-CI, it's happening on this line, because `u_num` is a float, not an integer:
```
======================================================================
ERROR: numpy.core.tests.test_function_base.TestFpspace.test_dtype
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python3.4/site-packages/numpy/core/function_base.py", line 15, in _index_deprecate
    i = operator.index(i)
TypeError: 'numpy.float64' object cannot be interpreted as an integer
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python3.4/site-packages/numpy/core/tests/test_function_base.py", line 348, in test_dtype
    y = fpspace(0, 6, dtype='float32')
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python3.4/site-packages/numpy/core/function_base.py", line 476, in fpspace
    u = linspace(u_start, u_stop, num=u_num, endpoint=True, dtype=dt)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python3.4/site-packages/numpy/core/function_base.py", line 101, in linspace
    num = _index_deprecate(num)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python3.4/site-packages/numpy/core/function_base.py", line 21, in _index_deprecate
    warnings.warn(msg, DeprecationWarning, stacklevel=stacklevel)
DeprecationWarning: object of type <class 'numpy.float64'> cannot be safely interpreted as an integer.
```

NumPy's test suite turns warnings issued internally into errors to ensure that we don't use deprecated behavior.

settle replied:
Thank you for pointing that out!  It behaves like gcc's -Werror and so an explicit cast to integer should do the trick.

shoyer commented:
I'm skeptical that this makes sense for NumPy.

I'm not an expert in floating-point numbers, but even aside from arguments like "subnormals", it's not at all obvious to me what this function does. It has 9 arguments merely for controlling the numbers it produces (aside from type related arguments), which is a very clear warning sign that it is trying to do too much.

I also don't see the common need for this functionality. In every case where I've needed scaling of this sort in the past, I have been happy to construct it directly from existing NumPy functionality.

Note that in general, we have a very high burden for new API since we try to avoid ever breaking backwards compatibility. Ultimately, adding new API requires consensus on the mailing-list, numpy-discussion@python.org.

settle commented:
You're entirely possible that this may be a bit to esoteric for Numpy.  It's subtly but significantly different than linspace between geomspace values due to the general normalization property of floating-point values.  I recently found it necessary for creating histogram bin edges that matched floating-point values as defined by IEEE.  Anyways, it was fun to extend my original overly simplistic version to be more robust and versatile like the Numpy spacing functions.

shoyer commented:
If you're excited to share this, it could certainly make for a good
stand-alone external package.

On Sat, Feb 24, 2018 at 10:27 PM settle <notifications@github.com> wrote:

> You're entirely possible that this may be a bit to esoteric for Numpy.
> It's subtly but significantly different than linspace between geomspace
> values due to the general normalization property of floating-point values.
> I recently found it necessary for creating histogram bin edges that matched
> floating-point values as defined by IEEE. Anyways, it was fun to extend my
> original overly simplistic version to be more robust and versatile like the
> Numpy spacing functions.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/10633#issuecomment-368287072>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABKS1qlSLlFhrWfjAtlSfTNLRGyvJMf2ks5tYP00gaJpZM4SLqDg>
> .
>

eric-wieser commented:
> I recently found it necessary for creating histogram bin edges that matched floating-point values as defined by IEEE

Can you elaborate on that?

settle commented:
I needed it for floating-point datatypes other than float32 and float64, and with the option to include/exclude subnormal values or extending the range of valid values in place of nan/inf as required by IEEE.  Unfortunately next_after wouldn't do the trick for my use case.
